File: web/src/lib/types/info/cpu.ts
import { z } from 'zod/v4';

export const CPUInfoHistoricalSchema = z.array(
    z.object({
        id: z.number().default(0),
        usage: z.number().default(0),
        createdAt: z.string().default('')
    })
);

export const ArchitectureSchema = z.enum([
    'amd64',
    '386',
    'arm',
    'arm64',
    'riscv64',
    'ppc64',
    'ppc64le',
    's390x',
    'wasm',
    'loong64'
]);

export const CPUInfoSchema = z.object({
    name: z.string().default('Unknown'),
    sockets: z.number().default(0),
    architecture: ArchitectureSchema.default('amd64'),
    physicalCores: z.number().default(0),
    threadsPerCore: z.number().default(0),
    logicalCores: z.number().default(0),
    family: z.number().default(0),
    model: z.number().default(0),
    features: z.array(z.string()).default([]),
    cacheLine: z.number().default(0),
    cache: z.object({
        l1d: z.number().default(0),
        l1i: z.number().default(0),
        l2: z.number().default(0),
        l3: z.number().default(0)
    }),
    frequency: z.number().default(0),
    usage: z.number().default(0)
});

export type Architecture = z.infer<typeof ArchitectureSchema>;
export type CPUInfo = z.infer<typeof CPUInfoSchema>;
export type CPUInfoHistorical = z.infer<typeof CPUInfoHistoricalSchema>;

File: internal/interfaces/services/info/cpu.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package infoServiceInterfaces

type Architecture string

const (
	ArchAMD64   Architecture = "amd64"
	Arch386     Architecture = "386"
	ArchARM     Architecture = "arm"
	ArchARM64   Architecture = "arm64"
	ArchRISCV64 Architecture = "riscv64"
	ArchPPC64   Architecture = "ppc64"
	ArchPPC64LE Architecture = "ppc64le"
	ArchS390X   Architecture = "s390x"
	ArchWASM    Architecture = "wasm"
	ArchLOONG64 Architecture = "loong64"
)

type CPUInfo struct {
	Name           string       `json:"name"`
	Architecture   Architecture `json:"architecture"`
	Sockets        int16        `json:"sockets"`
	PhysicalCores  int16        `json:"physicalCores"`
	ThreadsPerCore int16        `json:"threadsPerCore"`
	LogicalCores   int16        `json:"logicalCores"`
	Family         int16        `json:"family"`
	Model          int16        `json:"model"`
	Features       []string     `json:"features"`
	CacheLine      int16        `json:"cacheLine"`
	Cache          struct {
		L1D int16 `json:"l1d"`
		L1I int16 `json:"l1i"`
		L2  int16 `json:"l2"`
		L3  int16 `json:"l3"`
	} `json:"cache"`
	Frequency int64   `json:"frequency"`
	Usage     float64 `json:"usage"`
}
